src/libproc_macro/lib.rs

```diff
@@ -200,6 +201,78 @@ impl Span {
     pub fn call_site() -> Span {
         ::__internal::with_sess(|(_, mark)| Span(mark.expn_info().unwrap().call_site))
     }
+
+    /// The original source file into which this span points.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn source_file(&self) -> Option<SourceFile> {
+        // If this span's token was produced by a macro
+        if self.0.ctxt != SyntaxContext::empty() {
+            return None
+        }
+
+        let loc = __internal::lookup_char_pos(self.0.lo);
```

Comment:
Is it guaranteed that `lo` and `hi` point to the same file? In other words: are there spans which span multiple files? If there aren't, a comment explaining this would be nice for people reading the code and not knowing stuff about compiler internals (like me :stuck_out_tongue:). If there are, the doc comment should mention that the file of the *start* of the span is returned.

Reply:
You know, because of how the compiler basically sticks an entire crate's source code together, you could potentially have a span that covers multiple files if the macro invocation contains `mod` declarations and if you `glue()` the tokens from the individual spans together. 

So really `start()` and `end()` should probably return their own `SourceFile`'s, but that might be exposing too much of the compiler's implementation details. Either that or `glue()` should not allow spans with different source files to be glued together.

Reply:
If you glue the spans from the individual tokens together, that is. I'm on mobile for most of the day.

Reply:
What about having something like `BytePos` in `proc_macro`? Then `start()` and `end()` would return a `BytePos`. And there are methods like `BytePos::line()` and `BytePos::source_file()`. I think I'd like that API better; what do you think?